includes/class-wcsr-resource.php

```diff
@@ -187,24 +188,96 @@ public function get_days_active( $from_timestamp, $to_timestamp = null ) {
+
+			// If days based on time is only 1 but it was "across a day" we may need to adjust IF NOT accounted for already
+			if ( $days_by_time == 1 && ! self::is_on_same_day( $activation_time, $deactivation_time, $from_timestamp ) ) {
+
+				// handle situation if first activation crosses a day
+				if ( $i == 0 && ! self::is_on_same_day( $activation_time, $deactivation_time, $from_timestamp ) ) {
+					$days_active += 1;
+				}
+
+				// if this activation didn't start on the same day as previous activation it is safe to add an extra day
+				if ( $i !== 0 && ! self::is_on_same_day( $activation_time, $deactivation_times[ $i - 1 ], $from_timestamp ) ) {
+					$days_active += 1;
+				}
 			}
 		}
 
 		return $days_active;
 	}
 
 	/**
+	 * Prototype/demo conditional check for whether a timestamp is on the same "day" as another timestamp
+	 *
+	 * The catch is the "day" is not typical calendar day - it based on a 24 hour period from initial activation
+	 *
+	 * Initial activation could be the actual time it was first activated of the start of the period.
+	 *
+	 * Takes the activation/starting timestamp and gets the time our days start from.
+	 * Takes the timestamp we are checking if we are on the same day as and gets the date, previous days date and time
+	 * Works out whether the current day started on the same date as the one we are comparing with or a day earlier and works makes a time stampe fo when the day starts
+	 * Works out when the days ends
+	 * Check if our timestamp is with the start and end dates we have determined
+	 *
+	 * @param  int  $current_timestamp    [description]
+	 * @param  int  $compare_timestamp [description]
+	 * @param  int  $start_timestamp  [description]
+	 * @return boolean true on same day | false if not
+	 */
+	public static function is_on_same_day( $current_timestamp, $compare_timestamp, $start_timestamp ) {
+
+		$start_time = gmdate( 'H:i:s', $start_timestamp );
+		$start_hour = gmdate( 'H', $start_timestamp );
+		$start_min  = gmdate( 'i', $start_timestamp );
+		$start_sec  = gmdate( 's', $start_timestamp );
+
+		$compare_day  = gmdate( 'Y-m-d', $compare_timestamp );
+		$compare_day_previous  = gmdate( 'Y-m-d', strtotime( '-1 day', $compare_timestamp ) );
+		$compare_hour = gmdate( 'H', $compare_timestamp );
+		$compare_min  = gmdate( 'i', $compare_timestamp );
+		$compare_sec  = gmdate( 's', $compare_timestamp );
+
+		$start_of_the_day_date_time = $compare_day . ' ' . $start_time;
+		$start_of_the_day = strtotime( $start_of_the_day_date_time );
+
+		// Adjust for H:M:S being less
+		if ( $compare_hour < $start_hour || ( ( $compare_hour == $start_hour ) && ( $compare_min < $start_min ) ) || ( ( $compare_hour == $start_hour ) && ( $compare_min == $start_min ) && ( $compare_sec < $start_sec ) ) ) {
+			$start_of_the_day_date_time = $compare_day_previous . ' ' . $start_time;
+			$start_of_the_day = strtotime( $start_of_the_day_date_time );
+		}
+
+		$end_of_the_day = strtotime( '+1 day', $start_of_the_day );
```

Comment:
To determine `$start_of_the_day` and `$end_of_the_day`, I would personally use a loop to reduce the code and hopefully make it simpler.

The above 20 lines could be compacted to something like:

```php
for ( $end_of_the_day = $start_timestamp; $end_of_the_day < $compare_timestamp; $end_of_the_day + DAY_IN_SECONDS ) {
	// The loop controls take care of incrementing the end day (3rd expression) until the day after the compare date (2nd expression), but we also want to set the start date so we do that here using the current value of end day (which will be the start day in the final iteration as the 3rd expression in the loop hasn't run yet)
	$start_of_the_day = $end_of_the_day;
}
```

Or the slightly more verbose but easier to understand `do/while` equivalent:

```php
$end_of_the_day = $start_timestamp;

do {
	$end_of_the_day += DAY_IN_SECONDS;
} while ( $end_of_the_day < $compare_timestamp );

$start_of_the_day = $end_of_the_day - DAY_IN_SECONDS;
```

The problem with that one is that it assumes `$start_timestamp += DAY_IN_SECONDS` is always `< $compare_timestamp`, which the `for` loop doesn't. If that version is used, that condition would need to be checked at the outset of the function and enforced with an invalid arg exception.

(I haven't tested either of those, just for illustration purposes for now in case it's simpler for you to understand too).

Reply:
@thenbrent there were a few small changes I made to the `for loop` code you've given above.

1. change the condition to check `<=` - this was causing some tests to fail which were comparing the same timestamp as the start timestamp (see [these tests](https://github.com/Prospress/woocommerce-subscriptions-resource/blob/3513cacd8558ecc1a509658946cfa9aacd12b92d/tests/unit/class-wcsr-resource-test.php#L322-L351)). Before this change, these were the tests that were erroring/failing: https://cloudup.com/c75VWz2UfEu
2. use `+= DAY IN SECONDS` to stop the infinite loop

Reply:
Let me know if those changes look okay, or are going to cause terrible things to happen :)

Reply:
@mattallan haha they both sound like solid fixes. 👍